src/rgw/driver/rados/topic_migration.cc

```diff
@@ -0,0 +1,246 @@
+    ldpp_dout(dpp, 1) << "failed to read tenant, bucket name and marker from " << obj
+                      << " with: " << cpp_strerror(r)
+                      << " expected format pubsub.{tenant}.bucket.{name}/{marker}!" << dendl;
+    return r;
+  }
+
+  // migrate the notifications
+  rgw_pubsub_bucket_topics bucket_topics;
+  rgw_bucket rgw_bucket_info(tenant, bucket_name);
+  rgw_bucket_info.marker = marker;
+  rgw::sal::RadosBucket rados_bucket(driver, rgw_bucket_info);
+  RGWObjVersionTracker bucket_topics_objv;
+  r = rados_bucket.read_topics(bucket_topics, &bucket_topics_objv, y, dpp);
+  if (r == -ENOENT) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to read v1 bucket notifications from " << obj
+                      << " with: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  // in a for-loop that retries ECANCELED errors:
+  // {
+  // load the corresponding bucket by name
+  // break if marker doesn't match loaded bucket's
+  // break if RGW_ATTR_BUCKET_NOTIFICATION xattr already exists
+  // write RGW_ATTR_BUCKET_NOTIFICATION xattr
+  // }
+  std::unique_ptr<rgw::sal::Bucket> bucket;
+  do {
+    r = driver->load_bucket(dpp, rgw_bucket_info, &bucket, y);
+    if (r == ENOENT) {
+      break; // bucket is deleted, we should delete the v1 notification
+    }
+    if (r < 0) {
+      ldpp_dout(dpp, 1) << "failed to load the bucket from " << obj
+                        << " with: " << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    if (bucket->get_marker() != marker || bucket->get_attrs().contains(RGW_ATTR_BUCKET_NOTIFICATION)) {
+      break;
+    }
+
+    bufferlist bl;
+    bucket_topics.encode(bl);
+    bucket->get_attrs()[RGW_ATTR_BUCKET_NOTIFICATION] = std::move(bl);
+    r = bucket->put_info(dpp, false, real_time(), y);
+    if (r != ECANCELED && r < 0) {
+      ldpp_dout(dpp, 1) << "ERROR: failed writing bucket instance info: " << cpp_strerror(-r) << dendl;
+      return -r;
+    }
+  } while (r != ECANCELED);
+
+  // delete v1 notification obj with Bucket::remove_topics()
+  r = rados_bucket.remove_topics(&bucket_topics_objv, y, dpp);
+  if (r == -ECANCELED) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to remove migrated v1 bucket topics obj "
+                      << obj << " with: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  return r;
+}
+
+// migrate topics for a given tenant
+int migrate_topics(const DoutPrefixProvider* dpp, optional_yield y,
+                   rgw::sal::RadosStore* driver,
+                   const rgw_raw_obj& topics_obj)
+{
+  // parse tenant name out of topics_obj "pubsub.{tenant}"
+  std::string tenant;
+  const auto& topics_obj_oid = topics_obj.oid;
+  if (auto pos = topics_obj_oid.find(rgw::sal::pubsub_oid_prefix); pos != std::string::npos) {
+    tenant = topics_obj_oid.substr(pos, std::string(rgw::sal::pubsub_oid_prefix).size());
```

Comment:
given "pubsub.tenant", it looks like this would parse tenant = "pubsub."

i think we just want
```c++
  // parse tenant name out of topics_obj "pubsub.{tenant}"
  const std::string tenant = topics_obj.oid.substr(rgw::sal::pubsub_oid_prefix.size());
```

Reply:
Correct!